File: by-language/python/tests/test_trie.py
"""
Tests for Trie implementations.
"""

import pytest
from data_structures.trie import Trie, TrieMap, WordDictionary, CompressedTrie


class TestTrieBasics:
    """Test basic trie operations."""

    def test_init(self):
        """Test trie initialization."""
        trie = Trie()
        assert len(trie) == 0

    def test_insert_and_search(self):
        """Test basic insert and search."""
        trie = Trie()
        trie.insert("hello")

        assert trie.search("hello")
        assert not trie.search("hell")
        assert not trie.search("helloo")

    def test_insert_multiple(self):
        """Test inserting multiple words."""
        trie = Trie()
        words = ["cat", "car", "card", "care", "careful", "egg"]
        for word in words:
            trie.insert(word)

        assert len(trie) == 6
        for word in words:
            assert trie.search(word)

    def test_starts_with(self):
        """Test prefix checking."""
        trie = Trie()
        trie.insert("hello")
        trie.insert("help")
        trie.insert("world")

        assert trie.starts_with("hel")
        assert trie.starts_with("hello")
        assert trie.starts_with("wor")
        assert not trie.starts_with("abc")
        assert not trie.starts_with("helper")

    def test_insert_duplicate(self):
        """Test inserting duplicate word."""
        trie = Trie()
        trie.insert("test")
        trie.insert("test")

        assert len(trie) == 1
        assert trie.search("test")

    def test_empty_string(self):
        """Test handling empty string."""
        trie = Trie()
        trie.insert("")

        assert trie.search("")
        assert trie.starts_with("")

    def test_single_character(self):
        """Test single character words."""
        trie = Trie()
        trie.insert("a")
        trie.insert("b")

        assert trie.search("a")
        assert trie.search("b")
        assert not trie.search("c")


class TestTrieDelete:
    """Test trie deletion operations."""

    def test_delete_existing(self):
        """Test deleting existing word."""
        trie = Trie()
        trie.insert("hello")
        trie.insert("help")

        assert trie.delete("hello")
        assert not trie.search("hello")
        assert trie.search("help")
        assert len(trie) == 1

    def test_delete_nonexistent(self):
        """Test deleting non-existent word."""
        trie = Trie()
        trie.insert("hello")

        assert not trie.delete("world")
        assert len(trie) == 1

    def test_delete_prefix_word(self):
        """Test deleting word that is prefix of another."""
        trie = Trie()
        trie.insert("car")
        trie.insert("card")

        assert trie.delete("car")
        assert not trie.search("car")
        assert trie.search("card")

    def test_delete_longer_word(self):
        """Test deleting longer word with shared prefix."""
        trie = Trie()
        trie.insert("car")
        trie.insert("card")

        assert trie.delete("card")
        assert trie.search("car")
        assert not trie.search("card")

    def test_delete_all_words(self):
        """Test deleting all words."""
        trie = Trie()
        trie.insert("a")
        trie.insert("b")

        trie.delete("a")
        trie.delete("b")

        assert len(trie) == 0
        assert not trie.search("a")
        assert not trie.search("b")


class TestTriePrefixOperations:
    """Test prefix-based operations."""

    def test_get_words_with_prefix(self):
        """Test getting all words with prefix."""
        trie = Trie()
        trie.insert("cat")
        trie.insert("car")
        trie.insert("card")
        trie.insert("care")
        trie.insert("dog")

        result = trie.get_words_with_prefix("car")
        assert set(result) == {"car", "card", "care"}

    def test_get_words_with_prefix_no_match(self):
        """Test prefix with no matches."""
        trie = Trie()
        trie.insert("hello")

        assert trie.get_words_with_prefix("world") == []

    def test_get_words_with_empty_prefix(self):
        """Test empty prefix returns all words."""
        trie = Trie()
        trie.insert("a")
        trie.insert("b")
        trie.insert("c")

        result = trie.get_words_with_prefix("")
        assert set(result) == {"a", "b", "c"}

    def test_count_words_with_prefix(self):
        """Test counting words with prefix."""
        trie = Trie()
        trie.insert("test")
        trie.insert("testing")
        trie.insert("tested")
        trie.insert("tester")
        trie.insert("other")

        assert trie.count_words_with_prefix("test") == 4
        assert trie.count_words_with_prefix("testing") == 1
        assert trie.count_words_with_prefix("oth") == 1
        assert trie.count_words_with_prefix("xyz") == 0

    def test_autocomplete(self):
        """Test autocomplete functionality."""
        trie = Trie()
        words = ["apple", "application", "apply", "apt", "banana"]
        for word in words:
            trie.insert(word)

        result = trie.autocomplete("app", limit=2)
        assert len(result) == 2
        assert all(w.startswith("app") for w in result)

    def test_autocomplete_limit(self):
        """Test autocomplete respects limit."""
        trie = Trie()
        for i in range(100):
            trie.insert(f"word{i}")

        result = trie.autocomplete("word", limit=5)
        assert len(result) == 5


class TestTrieLongestCommonPrefix:
    """Test longest common prefix."""

    def test_common_prefix_exists(self):
        """Test finding common prefix."""
        trie = Trie()
        trie.insert("flower")
        trie.insert("flow")
        trie.insert("flight")

        assert trie.longest_common_prefix() == "fl"

    def test_no_common_prefix(self):
        """Test when no common prefix."""
        trie = Trie()
        trie.insert("dog")
        trie.insert("car")
        trie.insert("racecar")

        assert trie.longest_common_prefix() == ""

    def test_all_same_word(self):
        """Test all same words."""
        trie = Trie()
        trie.insert("test")

        assert trie.longest_common_prefix() == "test"

    def test_empty_trie(self):
        """Test empty trie."""
        trie = Trie()
        assert trie.longest_common_prefix() == ""


class TestTrieDunderMethods:
    """Test magic methods."""

    def test_len(self):
        """Test __len__."""
        trie = Trie()
        assert len(trie) == 0

        trie.insert("a")
        assert len(trie) == 1

        trie.insert("b")
        trie.insert("c")
        assert len(trie) == 3

    def test_contains(self):
        """Test __contains__."""
        trie = Trie()
        trie.insert("hello")

        assert "hello" in trie
        assert "world" not in trie

    def test_iter(self):
        """Test __iter__."""
        trie = Trie()
        words = ["cat", "dog", "bird"]
        for word in words:
            trie.insert(word)

        assert set(trie) == set(words)

    def test_repr(self):
        """Test __repr__."""
        trie = Trie()
        trie.insert("test")

        assert "Trie" in repr(trie)
        assert "1" in repr(trie)


class TestTrieMap:
    """Test TrieMap implementation."""

    def test_setitem_getitem(self):
        """Test setting and getting items."""
        tm = TrieMap()
        tm["hello"] = 1
        tm["world"] = 2

        assert tm["hello"] == 1
        assert tm["world"] == 2

    def test_getitem_missing(self):
        """Test getting missing key raises KeyError."""
        tm = TrieMap()
        with pytest.raises(KeyError):
            _ = tm["missing"]

    def test_contains(self):
        """Test membership check."""
        tm = TrieMap()
        tm["test"] = 1

        assert "test" in tm
        assert "missing" not in tm

    def test_get_with_default(self):
        """Test get with default value."""
        tm = TrieMap()
        tm["key"] = "value"

        assert tm.get("key") == "value"
        assert tm.get("missing") is None
        assert tm.get("missing", "default") == "default"

    def test_sum_with_prefix(self):
        """Test summing values with prefix."""
        tm = TrieMap()
        tm["apple"] = 10
        tm["application"] = 20
        tm["apply"] = 30
        tm["banana"] = 100

        assert tm.sum_with_prefix("app") == 60
        assert tm.sum_with_prefix("apple") == 10
        assert tm.sum_with_prefix("ban") == 100
        assert tm.sum_with_prefix("xyz") == 0

    def test_items_with_prefix(self):
        """Test getting items with prefix."""
        tm = TrieMap()
        tm["cat"] = 1
        tm["car"] = 2
        tm["card"] = 3
        tm["dog"] = 4

        items = tm.items_with_prefix("ca")
        assert set(items) == {("cat", 1), ("car", 2), ("card", 3)}

    def test_len(self):
        """Test length."""
        tm = TrieMap()
        assert len(tm) == 0

        tm["a"] = 1
        tm["b"] = 2
        assert len(tm) == 2

        # Update existing key
        tm["a"] = 10
        assert len(tm) == 2


class TestWordDictionary:
    """Test WordDictionary with wildcard support."""

    def test_basic_operations(self):
        """Test basic add and search."""
        wd = WordDictionary()
        wd.add_word("bad")
        wd.add_word("dad")
        wd.add_word("mad")

        assert wd.search("bad")
        assert wd.search("dad")
        assert not wd.search("pad")

    def test_wildcard_single(self):
        """Test single wildcard."""
        wd = WordDictionary()
        wd.add_word("bad")
        wd.add_word("dad")

        assert wd.search("b.d")
        assert wd.search("d.d")
        assert wd.search(".ad")
        assert not wd.search(".od")

    def test_wildcard_multiple(self):
        """Test multiple wildcards."""
        wd = WordDictionary()
        wd.add_word("bad")

        assert wd.search("...")
        assert wd.search("b..")
        assert wd.search(".a.")
        assert wd.search("..d")
        assert not wd.search("....")

    def test_wildcard_all(self):
        """Test all wildcards."""
        wd = WordDictionary()
        wd.add_word("abc")
        wd.add_word("xyz")

        assert wd.search("...")

    def test_no_match(self):
        """Test no match cases."""
        wd = WordDictionary()
        wd.add_word("hello")

        assert not wd.search("hell")
        assert not wd.search("helloo")
        assert not wd.search("h.....")  # 6 dots - too long
        assert wd.search("h....")  # 5 chars should match "hello"


class TestCompressedTrie:
    """Test compressed trie (radix tree)."""

    def test_insert_and_search(self):
        """Test basic insert and search."""
        ct = CompressedTrie()
        ct.insert("test")
        ct.insert("testing")
        ct.insert("tested")

        assert ct.search("test")
        assert ct.search("testing")
        assert ct.search("tested")
        assert not ct.search("tes")

    def test_starts_with(self):
        """Test prefix checking."""
        ct = CompressedTrie()
        ct.insert("hello")
        ct.insert("help")

        assert ct.starts_with("hel")
        assert ct.starts_with("hello")
        assert not ct.starts_with("world")

    def test_shared_prefix(self):
        """Test words with shared prefix."""
        ct = CompressedTrie()
        ct.insert("romane")
        ct.insert("romanus")
        ct.insert("romulus")
        ct.insert("rubens")

        assert ct.search("romane")
        assert ct.search("romanus")
        assert ct.search("romulus")
        assert ct.search("rubens")
        assert not ct.search("rom")
        assert not ct.search("roman")

    def test_len(self):
        """Test length."""
        ct = CompressedTrie()
        assert len(ct) == 0

        ct.insert("a")
        ct.insert("ab")
        ct.insert("abc")
        assert len(ct) == 3

    def test_contains(self):
        """Test membership."""
        ct = CompressedTrie()
        ct.insert("test")

        assert "test" in ct
        assert "testing" not in ct

    def test_edge_splitting(self):
        """Test edge splitting when inserting."""
        ct = CompressedTrie()
        ct.insert("test")
        ct.insert("team")  # Should split at 'te'

        assert ct.search("test")
        assert ct.search("team")

    def test_exact_prefix_insert(self):
        """Test inserting exact prefix."""
        ct = CompressedTrie()
        ct.insert("testing")
        ct.insert("test")  # Insert prefix of existing

        assert ct.search("test")
        assert ct.search("testing")


class TestTrieEdgeCases:
    """Test edge cases."""

    def test_unicode_characters(self):
        """Test unicode support."""
        trie = Trie()
        trie.insert("hello")
        trie.insert("helloworld")

        assert trie.search("hello")
        assert trie.search("helloworld")

    def test_long_words(self):
        """Test very long words."""
        trie = Trie()
        long_word = "a" * 1000
        trie.insert(long_word)

        assert trie.search(long_word)
        assert not trie.search(long_word[:-1])

    def test_many_similar_prefixes(self):
        """Test many words with similar prefixes."""
        trie = Trie()
        for i in range(100):
            trie.insert(f"prefix{i}")

        assert len(trie) == 100
        assert trie.starts_with("prefix")
        assert trie.count_words_with_prefix("prefix") == 100

    def test_special_characters(self):
        """Test special characters in words."""
        trie = Trie()
        trie.insert("hello-world")
        trie.insert("hello_world")
        trie.insert("hello.world")

        assert trie.search("hello-world")
        assert trie.search("hello_world")
        assert trie.search("hello.world")

File: by-language/c/tests/test_trie.c
/**
 * Tests for Trie (Prefix Tree)
 */

#include "test_framework.h"
#include "../data-structures/trie.h"
#include <stdlib.h>
#include <string.h>

/* ============== Creation Tests ============== */

TEST(trie_create_empty) {
    Trie *trie = trie_create();
    ASSERT_NOT_NULL(trie);
    ASSERT_EQ(0, trie_word_count(trie));
    ASSERT_TRUE(trie_is_empty(trie));
    trie_destroy(trie);
}

/* ============== Insert Tests ============== */

TEST(trie_insert_single) {
    Trie *trie = trie_create();
    ASSERT_TRUE(trie_insert(trie, "hello"));
    ASSERT_EQ(1, trie_word_count(trie));
    ASSERT_TRUE(trie_search(trie, "hello"));
    trie_destroy(trie);
}

TEST(trie_insert_multiple) {
    Trie *trie = trie_create();
    ASSERT_TRUE(trie_insert(trie, "hello"));
    ASSERT_TRUE(trie_insert(trie, "world"));
    ASSERT_TRUE(trie_insert(trie, "help"));
    ASSERT_EQ(3, trie_word_count(trie));
    trie_destroy(trie);
}

TEST(trie_insert_duplicate) {
    Trie *trie = trie_create();
    ASSERT_TRUE(trie_insert(trie, "hello"));
    ASSERT_TRUE(trie_insert(trie, "hello")); /* Duplicate */
    ASSERT_EQ(1, trie_word_count(trie));
    trie_destroy(trie);
}

TEST(trie_insert_prefix) {
    Trie *trie = trie_create();
    ASSERT_TRUE(trie_insert(trie, "app"));
    ASSERT_TRUE(trie_insert(trie, "apple"));
    ASSERT_EQ(2, trie_word_count(trie));
    ASSERT_TRUE(trie_search(trie, "app"));
    ASSERT_TRUE(trie_search(trie, "apple"));
    trie_destroy(trie);
}

/* ============== Search Tests ============== */

TEST(trie_search_found) {
    Trie *trie = trie_create();
    trie_insert(trie, "hello");
    trie_insert(trie, "world");
    ASSERT_TRUE(trie_search(trie, "hello"));
    ASSERT_TRUE(trie_search(trie, "world"));
    trie_destroy(trie);
}

TEST(trie_search_not_found) {
    Trie *trie = trie_create();
    trie_insert(trie, "hello");
    ASSERT_FALSE(trie_search(trie, "hell"));
    ASSERT_FALSE(trie_search(trie, "helloworld"));
    ASSERT_FALSE(trie_search(trie, "world"));
    trie_destroy(trie);
}

TEST(trie_search_prefix_not_word) {
    Trie *trie = trie_create();
    trie_insert(trie, "apple");
    ASSERT_FALSE(trie_search(trie, "app")); /* Prefix but not inserted */
    ASSERT_TRUE(trie_search(trie, "apple"));
    trie_destroy(trie);
}

/* ============== Starts With Tests ============== */

TEST(trie_starts_with_true) {
    Trie *trie = trie_create();
    trie_insert(trie, "apple");
    trie_insert(trie, "application");
    ASSERT_TRUE(trie_starts_with(trie, "app"));
    ASSERT_TRUE(trie_starts_with(trie, "appl"));
    ASSERT_TRUE(trie_starts_with(trie, "apple"));
    trie_destroy(trie);
}

TEST(trie_starts_with_false) {
    Trie *trie = trie_create();
    trie_insert(trie, "apple");
    ASSERT_FALSE(trie_starts_with(trie, "apl"));
    ASSERT_FALSE(trie_starts_with(trie, "banana"));
    trie_destroy(trie);
}

/* ============== Delete Tests ============== */

TEST(trie_delete_single) {
    Trie *trie = trie_create();
    trie_insert(trie, "hello");
    ASSERT_TRUE(trie_delete(trie, "hello"));
    ASSERT_FALSE(trie_search(trie, "hello"));
    ASSERT_EQ(0, trie_word_count(trie));
    trie_destroy(trie);
}

TEST(trie_delete_preserves_prefix) {
    Trie *trie = trie_create();
    trie_insert(trie, "app");
    trie_insert(trie, "apple");
    ASSERT_TRUE(trie_delete(trie, "app"));
    ASSERT_FALSE(trie_search(trie, "app"));
    ASSERT_TRUE(trie_search(trie, "apple"));
    ASSERT_EQ(1, trie_word_count(trie));
    trie_destroy(trie);
}

TEST(trie_delete_not_found) {
    Trie *trie = trie_create();
    trie_insert(trie, "hello");
    ASSERT_FALSE(trie_delete(trie, "world"));
    ASSERT_EQ(1, trie_word_count(trie));
    trie_destroy(trie);
}

/* ============== Count Prefix Tests ============== */

TEST(trie_count_prefix_basic) {
    Trie *trie = trie_create();
    trie_insert(trie, "apple");
    trie_insert(trie, "application");
    trie_insert(trie, "apply");
    trie_insert(trie, "banana");
    ASSERT_EQ(3, trie_count_prefix(trie, "app"));
    ASSERT_EQ(1, trie_count_prefix(trie, "ban"));
    ASSERT_EQ(0, trie_count_prefix(trie, "xyz"));
    trie_destroy(trie);
}

/* ============== Autocomplete Tests ============== */

TEST(trie_autocomplete_basic) {
    Trie *trie = trie_create();
    trie_insert(trie, "apple");
    trie_insert(trie, "application");
    trie_insert(trie, "apply");
    trie_insert(trie, "banana");

    char *words[10];
    int count = trie_autocomplete(trie, "app", words, 10);
    ASSERT_EQ(3, count);

    /* Free allocated strings */
    for (int i = 0; i < count; i++) {
        free(words[i]);
    }
    trie_destroy(trie);
}

TEST(trie_autocomplete_no_results) {
    Trie *trie = trie_create();
    trie_insert(trie, "apple");

    char *words[10];
    int count = trie_autocomplete(trie, "xyz", words, 10);
    ASSERT_EQ(0, count);
    trie_destroy(trie);
}

/* ============== Get All Words Tests ============== */

TEST(trie_get_all_words_basic) {
    Trie *trie = trie_create();
    trie_insert(trie, "cat");
    trie_insert(trie, "car");
    trie_insert(trie, "dog");

    char *words[10];
    int count = trie_get_all_words(trie, words, 10);
    ASSERT_EQ(3, count);

    for (int i = 0; i < count; i++) {
        free(words[i]);
    }
    trie_destroy(trie);
}

/* ============== Wildcard Search Tests ============== */

TEST(trie_wildcard_search_found) {
    Trie *trie = trie_create();
    trie_insert(trie, "bad");
    trie_insert(trie, "dad");
    trie_insert(trie, "mad");

    ASSERT_TRUE(trie_search_wildcard(trie, ".ad"));
    ASSERT_TRUE(trie_search_wildcard(trie, "b.."));
    ASSERT_TRUE(trie_search_wildcard(trie, "..."));
    trie_destroy(trie);
}

TEST(trie_wildcard_search_not_found) {
    Trie *trie = trie_create();
    trie_insert(trie, "bad");

    ASSERT_FALSE(trie_search_wildcard(trie, "...."));
    ASSERT_FALSE(trie_search_wildcard(trie, "b.x"));
    trie_destroy(trie);
}

/* ============== Longest Common Prefix Tests ============== */

TEST(trie_lcp_basic) {
    Trie *trie = trie_create();
    trie_insert(trie, "flower");
    trie_insert(trie, "flow");
    trie_insert(trie, "flight");

    char result[20];
    int len = trie_longest_common_prefix(trie, result, 20);
    ASSERT_EQ(2, len);
    ASSERT_EQ(0, strcmp("fl", result));
    trie_destroy(trie);
}

TEST(trie_lcp_no_common) {
    Trie *trie = trie_create();
    trie_insert(trie, "dog");
    trie_insert(trie, "cat");

    char result[20];
    int len = trie_longest_common_prefix(trie, result, 20);
    ASSERT_EQ(0, len);
    trie_destroy(trie);
}

/* ============== Shortest Unique Prefix Tests ============== */

TEST(trie_shortest_unique_prefix_basic) {
    Trie *trie = trie_create();
    trie_insert(trie, "zebra");
    trie_insert(trie, "dog");
    trie_insert(trie, "duck");
    trie_insert(trie, "dove");

    char result[20];
    int len = trie_shortest_unique_prefix(trie, "zebra", result, 20);
    ASSERT_EQ(1, len);
    ASSERT_EQ(0, strcmp("z", result));

    len = trie_shortest_unique_prefix(trie, "dog", result, 20);
    ASSERT_EQ(3, len);
    ASSERT_EQ(0, strcmp("dog", result));
    trie_destroy(trie);
}

/* ============== Clear Tests ============== */

TEST(trie_clear_test) {
    Trie *trie = trie_create();
    trie_insert(trie, "hello");
    trie_insert(trie, "world");
    ASSERT_EQ(2, trie_word_count(trie));

    trie_clear(trie);
    ASSERT_EQ(0, trie_word_count(trie));
    ASSERT_TRUE(trie_is_empty(trie));
    ASSERT_FALSE(trie_search(trie, "hello"));
    trie_destroy(trie);
}

/* ============== NULL Safety Tests ============== */

TEST(trie_null_safety) {
    ASSERT_FALSE(trie_insert(NULL, "test"));
    ASSERT_FALSE(trie_search(NULL, "test"));
    ASSERT_FALSE(trie_delete(NULL, "test"));
    ASSERT_FALSE(trie_starts_with(NULL, "test"));
    ASSERT_EQ(0, trie_count_prefix(NULL, "test"));
    ASSERT_EQ(0, trie_word_count(NULL));
    ASSERT_TRUE(trie_is_empty(NULL));
}

/* ============== Case Sensitivity Tests ============== */

TEST(trie_case_insensitive) {
    Trie *trie = trie_create();
    trie_insert(trie, "Hello");
    ASSERT_TRUE(trie_search(trie, "hello"));
    ASSERT_TRUE(trie_search(trie, "HELLO"));
    ASSERT_TRUE(trie_search(trie, "HeLLo"));
    trie_destroy(trie);
}

/* Main test runner */
int main(void) {
    TEST_SUITE_START("Trie (Prefix Tree)");

    /* Creation */
    RUN_TEST(trie_create_empty);

    /* Insert */
    RUN_TEST(trie_insert_single);
    RUN_TEST(trie_insert_multiple);
    RUN_TEST(trie_insert_duplicate);
    RUN_TEST(trie_insert_prefix);

    /* Search */
    RUN_TEST(trie_search_found);
    RUN_TEST(trie_search_not_found);
    RUN_TEST(trie_search_prefix_not_word);

    /* Starts With */
    RUN_TEST(trie_starts_with_true);
    RUN_TEST(trie_starts_with_false);

    /* Delete */
    RUN_TEST(trie_delete_single);
    RUN_TEST(trie_delete_preserves_prefix);
    RUN_TEST(trie_delete_not_found);

    /* Count Prefix */
    RUN_TEST(trie_count_prefix_basic);

    /* Autocomplete */
    RUN_TEST(trie_autocomplete_basic);
    RUN_TEST(trie_autocomplete_no_results);

    /* Get All Words */
    RUN_TEST(trie_get_all_words_basic);

    /* Wildcard Search */
    RUN_TEST(trie_wildcard_search_found);
    RUN_TEST(trie_wildcard_search_not_found);

    /* Longest Common Prefix */
    RUN_TEST(trie_lcp_basic);
    RUN_TEST(trie_lcp_no_common);

    /* Shortest Unique Prefix */
    RUN_TEST(trie_shortest_unique_prefix_basic);

    /* Clear */
    RUN_TEST(trie_clear_test);

    /* NULL Safety */
    RUN_TEST(trie_null_safety);

    /* Case */
    RUN_TEST(trie_case_insensitive);

    TEST_SUITE_END();

    return GET_FAILED_COUNT() > 0 ? 1 : 0;
}
